internal/jobcontainers/jobcontainer.go

```diff
@@ -0,0 +1,393 @@
+	go container.waitBackground(ctx)
+	return container, nil
+}
+
+// CreateProcess creates a process on the host, starts it, adds it to the containers
+// job object and then waits for exit.
+func (c *JobContainer) CreateProcess(ctx context.Context, config interface{}) (_ cow.Process, err error) {
+	conf, ok := config.(*hcsschema.ProcessParameters)
+	if !ok {
+		return nil, errors.New("unsupported process config passed in")
+	}
+
+	path, args := seperateArgs(conf.CommandLine)
+	absPath, err := findExecutable(path, c.sandboxMount)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find executable")
+	}
+
+	token, err := processToken(conf.User)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create process token")
+	}
+
+	cmd := &exec.Cmd{
+		Env:  os.Environ(),
+		Dir:  c.sandboxMount,
+		Path: absPath,
+		Args: args,
+		SysProcAttr: &syscall.SysProcAttr{
+			CreationFlags: windows.CREATE_NEW_PROCESS_GROUP,
+			Token:         syscall.Token(token),
+		},
+	}
+
+	process := newProcess(cmd)
+
+	var (
+		stdin  io.WriteCloser
+		stdout io.ReadCloser
+		stderr io.ReadCloser
+	)
+	// Create process pipes if asked for.
+	if conf.CreateStdInPipe {
+		stdin, err = process.cmd.StdinPipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdin pipe")
+		}
+		process.stdin = stdin
+	}
+
+	if conf.CreateStdOutPipe {
+		stdout, err = process.cmd.StdoutPipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdout pipe")
+		}
+		process.stdout = stdout
+	}
+
+	if conf.CreateStdErrPipe {
+		stderr, err = process.cmd.StderrPipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stderr pipe")
+		}
+		process.stderr = stderr
+	}
+
+	defer func() {
+		if err != nil {
+			process.Close()
+			c.Release(ctx)
+		}
+	}()
+
+	if err = process.Start(); err != nil {
+		return nil, errors.Wrap(err, "failed to start host process")
+	}
+
+	if err = c.job.Assign(uint32(process.Pid())); err != nil {
+		return nil, errors.Wrap(err, "failed to assign process to job object")
+	}
+
+	// Assign the first process made as the init process of the container.
+	c.init.initDoOnce.Do(func() {
+		c.init.proc = process
+		close(c.init.initBlock)
+	})
+
+	// Wait for process exit
+	go c.pollJobMsgs(ctx)
+	go process.waitBackground(ctx)
+	return process, nil
+}
+
+// Release unmounts all of the container layers. Safe to call multiple times, if no storage
+// is mounted this call will just return nil.
+func (c *JobContainer) Release(ctx context.Context) error {
+	if c.sandboxMount != "" {
+		if err := removeSandboxMountPoint(ctx, c.sandboxMount); err != nil {
+			log.G(ctx).WithFields(logrus.Fields{
+				"id":   c.id,
+				"path": c.sandboxMount,
+			}).Warn("failed to remove sandbox volume mount")
+			return errors.Wrap(err, "failed to remove sandbox volume mount path")
+		}
+		if err := layers.UnmountContainerLayers(ctx, c.spec.Windows.LayerFolders, "", nil, layers.UnmountOperationAll); err != nil {
```

Comment:
Wouldn't this cause an error if a Root path was specified on the spec?

Reply:
If the root path was specified on the spec given to Create, we would not hit the path that calls `MountContainerLayers` in [`mountLayers`](https://github.com/microsoft/hcsshim/blob/1878f1e96d79c356428b5a1898b58798e7254efd/internal/jobcontainers/storage.go#L47), but since a root path does exist, we will continue in `Create` to [set the container's `sandboxMount`](https://github.com/microsoft/hcsshim/blob/1878f1e96d79c356428b5a1898b58798e7254efd/internal/jobcontainers/jobcontainer.go#L141). This could be problematic now on release, since if the `sandboxMount` is not empty we [attempt to unmount the layers](https://github.com/microsoft/hcsshim/blob/1878f1e96d79c356428b5a1898b58798e7254efd/internal/jobcontainers/jobcontainer.go#L252) that would have otherwise been mounted in `mountLayers` IF a root path had not been specified on the initial spec. I may have missed something here so maybe this isn't an issue, but worth looking into. Do we know if this scenario is even possible though?

Reply:
Root.Path doesn't get populated for WCOW AFAIK as we Activate and Prepare the layer in the shim itself from the layer folders. If instead we did this setup earlier and we were given the root path on the spec then nothing really changes either. If we were given a root path that means the layers have been activated already so theres no need to call `MountContainerLayers`. 

Inside `UnmountContainerLayers` `UnprepareLayer` would be called which will remove the wcifs filter from the volume, and then `DeactivateLayer` just unmounts the vhd so either way things should still function.